File: src/prbm_stackqueue/BridgePassingTruck.java
package prbm_stackqueue;

import java.util.LinkedList;
import java.util.Queue;

public class BridgePassingTruck {

	public static void main(String[] args) {
		System.out.println(
				solution(2, 
						10, 
						new int[]{1,4,5,6})
				);

	}

    public static int solution(int bridge_length, int weight, int[] truck_weights) {
    	// 트럭 1대가 지나는 시간은 무조건 bridge_length+1
    	// 분명 Queue쓰는 문젠데, 정해진 순서로만 건너가는게 포인트
    	Queue<Truck> queue = new LinkedList<Truck>();
    	
    	int totalTruck = truck_weights.length;    	
    	int sec=0;
    	int beforeIndex = 0;
    	int afterIndex = 0;
    	int onBridge=0;
    	
    	while(beforeIndex<totalTruck) {
    		++sec;
    		for(Truck truck : queue) {
    			truck.sec+=1;
    		}
    		if(queue.peek()!=null && queue.peek().sec>bridge_length) {
    			queue.remove();
    			onBridge-=truck_weights[afterIndex];
    			++afterIndex;
    		}
    		if(onBridge+truck_weights[beforeIndex] <= weight) {
    			queue.offer(new Truck());
    			onBridge+=truck_weights[beforeIndex];
    			++beforeIndex;
    		}
    	}    	    
    	return sec+bridge_length;
    }

}

class Truck {
	int sec=1;
}

File: src/py/pg/prbm_stackqueue/BridgePassingTruck.py
import queue

class Truck:
    def __init__(self, weight):
        self.location = 1
        self.weight = weight

    def oneSec(self):
        self.location += 1

def solution(bridge_length, weight, truck_weights):
    sec = 0
    truckOnBridgeQueue = list()
    weightOnBridge = 0
    
    while len(truck_weights) > 0 :
        for truck in truckOnBridgeQueue :
            truck.oneSec()
            
        if len(truckOnBridgeQueue)>0 and truckOnBridgeQueue[0].location > bridge_length :
            weightOnBridge -= truckOnBridgeQueue[0].weight                
            del truckOnBridgeQueue[0]

        if weightOnBridge+truck_weights[0] <= weight :
            truckOnBridgeQueue.append(Truck(truck_weights[0]))
            weightOnBridge += truck_weights[0]
            del truck_weights[0]        
        sec += 1
        
    return sec+bridge_length



# 틀렸던 부분
def solutionWrong(bridge_length, weight, truck_weights):
    sec = 0
    truckOnBridgeQueue = list()
    weightOnBridge = 0
    
    while len(truck_weights) > 0 :
        for truck in truckOnBridgeQueue :
            truck.oneSec()            
            if truck.location > bridge_length :
                weightOnBridge -= truckOnBridgeQueue[0].weight                
                del truckOnBridgeQueue[0] # for loop 순회하던 중 큐의 원소를 지워버려서 제대로 순회 못하는 문제가 발생했었다.
                # loop 순회 중에는 loop의 길이에 변화를 주는 행동은 하지 말것

        if weightOnBridge+truck_weights[0] <= weight :
            truckOnBridgeQueue.append(Truck(truck_weights[0]))
            weightOnBridge += truck_weights[0]
            del truck_weights[0]        
        sec += 1
        
    return sec+bridge_length